compositor_common/src/scene/validation/constraints/input_count.rs

```diff
@@ -0,0 +1,56 @@
+use std::rc::Rc;
+
+use serde::{Deserialize, Serialize};
+
+use crate::{error::UnsatisfiedConstraintsError, scene::NodeSpec};
+
+#[derive(Debug, PartialEq, Eq, Clone, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum InputsCountConstraint {
+    Exactly(u32),
+    // TODO validate that lower bound <= upper bound
+    Range { lower_bound: u32, upper_bound: u32 },
+}
+
+impl InputsCountConstraint {
+    pub fn check(&self, node_spec: &NodeSpec) -> Result<(), UnsatisfiedConstraintsError> {
+        let defined_input_pads_count = node_spec.input_pads.len() as u32;
+        let is_valid = match self {
+            InputsCountConstraint::Exactly(expected_inputs_count) => {
+                defined_input_pads_count == *expected_inputs_count
+            }
+            InputsCountConstraint::Range {
+                lower_bound,
+                upper_bound,
+            } => *lower_bound < defined_input_pads_count && defined_input_pads_count < *upper_bound,
+        };
+
+        if is_valid {
+            Ok(())
+        } else {
+            Err(UnsatisfiedConstraintsError::InvalidInputsCount {
+                node_id: node_spec.node_id.clone(),
+                identification_name: node_spec.identification_name(),
+                input_count_constrain: self.clone(),
+                defined_input_pads_count,
+            })
+        }
+    }
+
+    pub fn required_inputs_message(&self) -> Rc<str> {
```

Comment:
methods that return string that will be concatenated with sth else are very sketch, it's very easy for  that to break, if you need sth like that it should be as private as possible and close to the place where it's used. Yes, there is sth similar in registry implementation, but I would like to get rid of that too.